src/box/lua/net_box.c

```diff
@@ -100,19 +113,125 @@ enum netbox_method {
+};
+
+static const char *netbox_state_str[] = {
+	[NETBOX_INITIAL]         = "initial",
+	[NETBOX_AUTH]            = "auth",
+	[NETBOX_FETCH_SCHEMA]    = "fetch_schema",
+	[NETBOX_ACTIVE]          = "active",
+	[NETBOX_ERROR]           = "error",
+	[NETBOX_ERROR_RECONNECT] = "error_reconnect",
+	[NETBOX_CLOSED]          = "closed",
+};
+
+struct netbox_options {
+	/** Remote server host and port. */
+	char *host;
+	char *service;
+	/** User credentials. */
+	char *user;
+	char *password;
+	/**
+	 * Lua reference to the transport callback function
+	 * (see the comment to netbox_transport).
+	 */
+	int callback_ref;
+	/** connect() timeout, in seconds. */
+	double connect_timeout;
+	/**
+	 * Timeout to wait after a connection failure before trying to
+	 * reconnect, in seconds. Reconnect is disabled if it's 0.
+	 */
+	double reconnect_after;
+};
+
+/**
+ * Basically, *transport* is a TCP connection speaking the Tarantool
+ * network protocol (IPROTO). This is a low-level interface.
+ * Primary features:
+ *  * concurrent perform_request()-s benefit from multiplexing
+ *    support in the protocol;
+ *  * schema-aware - snoops responses and initiates schema reload
+ *    when a response has a new schema version;
+ *  * delivers transport events via the callback.
+ *
+ * Transport state machine:
+ *
+ * State machine starts in 'initial' state. Start method
+ * spawns a worker fiber, which will establish a connection.
+ * Stop method sets the state to 'closed' and kills the worker.
+ * If the transport is already in 'error' state stop() does
+ * nothing.
+ *
+ * State chart:
+ *
+ *  initial -> auth -> fetch_schema <-> active
+ *
+ *  (any state, on error) -> error_reconnect -> auth -> ...
+ *                                           \
+ *                                            -> error
+ *  (any state, but 'error') -> closed
+ *
+ *
+ * State change events can be delivered to the transport user via
+ * the optional 'callback' argument:
+ *
+ * The callback functions needs to have the following signature:
+ *
+ *  callback(event_name, ...)
+ *
+ * The following events are delivered, with arguments:
+ *
+ *  'state_changed', state, error
+ *  'handshake', greeting, version, features
+ *  'did_fetch_schema', schema_version, spaces, indices, collations
+ *  'event', key, value
+ */
 struct netbox_transport {
+	/** Connection options. Not modified after initialization. */
+	struct netbox_options opts;
+	/** Greeting received from the remote host. */
+	struct greeting greeting;
+	/** Connection state. */
+	enum netbox_state state;
+	/**
+	 * The connection is closing. No new requests are allowed.
+	 * The connection will be closed as soon as all pending requests
+	 * have been sent.
+	 */
+	bool is_closing;
+	/** Error that caused the last connection failure or NULL. */
+	struct error *last_error;
+	/** Fiber doing I/O and dispatching responses. */
+	struct fiber *worker;
+	/** Lua thread used by the worker fiber or NULL. */
+	struct lua_State *L;
```

Comment:
Agree. The transport callback may need a state (e.g. for running on_connect triggers). Better set `storage.lua.stack` so that it can reuse the existing state instead of creating a new one. Done. Thanks!